proxy/src/main/java/com/wavefront/agent/TenantInfo.java

```diff
@@ -0,0 +1,230 @@
+import static com.wavefront.agent.ProxyConfigDef.cspBaseUrl;
+import static java.lang.String.format;
+import static javax.ws.rs.core.Response.Status.Family.SUCCESSFUL;
+
+import com.google.common.net.HttpHeaders;
+import com.wavefront.common.NamedThreadFactory;
+import java.util.Base64;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nonnull;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.core.Form;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder;
+
+/**
+ * The class for keeping tenant required information token, server.
+ *
+ * @author Norayr Chaparyan (nchaparyan@vmware.com).
+ */
+public class TenantInfo {
+  private static final String GET_CSP_ACCESS_TOKEN_ERROR_MESSAGE =
+      "Failed to get access token from CSP.";
+  private static final ScheduledExecutorService executor =
+      Executors.newScheduledThreadPool(1, new NamedThreadFactory("token-configuration"));
+  private String bearerToken;
+  private final String server;
+
+  public TenantInfo(
+      @Nonnull final String token,
+      @Nonnull final String server,
+      ProxyConfig.ProxyAuthMethod proxyAuthMethod) {
+    this.server = server;
+    switch (proxyAuthMethod) {
+      case CSP_API_TOKEN:
+        loadAccessTokenByAPIToken(token, 0);
+        break;
+      case WAVEFRONT_API_TOKEN:
+        bearerToken = token;
+        break;
+      default:
+        throw new IllegalStateException(
+            "CSP_API_TOKEN or WAVEFRONT_API_TOKEN should be provided as proxy auth method.");
+    }
+  }
+
+  public TenantInfo(
+      @Nonnull final String clientId,
+      @Nonnull final String clientSecret,
+      @Nonnull final String orgId,
+      @Nonnull final String server) {
+    this.server = server;
+    loadAccessTokenByClientCredentials(clientId, clientSecret, orgId, 0);
+  }
+
+  @Nonnull
+  public String getServer() {
+    return server;
+  }
+
+  @Nonnull
+  public String getToken() {
+    return bearerToken;
+  }
+
+  /**
+   * When the CSP access token has expired, obtain it using the CSP api token.
+   *
+   * @param apiToken the CSP api token.
+   * @param failedRequestsCount the number of unsuccessful requests.
+   * @throws RuntimeException when CSP is down or wrong CSP api token.
+   */
+  private void loadAccessTokenByAPIToken(@Nonnull final String apiToken, int failedRequestsCount) {
+
+    Form requestBody = new Form();
+    requestBody.param("grant_type", "api_token");
+    requestBody.param("api_token", apiToken);
+
+    Response response =
+        new ResteasyClientBuilder()
+            .build()
+            .target(format("%s/csp/gateway/am/api/auth/api-tokens/authorize", cspBaseUrl))
+            .request()
+            .post(Entity.form(requestBody));
+
+    if (response.getStatusInfo().getFamily() != SUCCESSFUL) {
+      if (failedRequestsCount < 3) {
```

Comment:
why retry only 3 times ?
I think this should be implemented on a simpler way..
If we get token schedule in `getExpiresIn` if not schedule in 1 min.